fix(is_installed): Coerce user-specified version to `numeric_version()`

gadenbuie opened the pull request:
In light of the recent CRAN changes around version handling, `is_installed()` currently may introduce a hard-to-discover instance of incorrect version specification by comparing a numeric value against a package version.

We could required all uses of `is_installed()` to diligently ensure the `version` argument is a character value, or we can coerce the version to `numeric_version()` prior to the comparison. This PR implements the latter so that we can update any statically imported `is_installed()` and know the issue is mitigated.

I chose `numeric_version()` over `as.character()`, `package_version()`, or `as.package_version()` because it's the most robust to all package version variants.

``` r
version_dbl <- 2.0
version_chr <- "2.0"
version_pkg <- "2-0"

version_dbl |> package_version()
#> Error: invalid version specification '2'
version_chr |> package_version()
#> [1] '2.0'
version_pkg |> package_version()
#> [1] '2.0'

version_dbl |> as.character() |> package_version()
#> Error: invalid version specification '2'
version_chr |> as.character() |> package_version()
#> [1] '2.0'
version_pkg |> as.character() |> package_version()
#> [1] '2.0'

version_dbl |> numeric_version()
#> [1] '2'
version_chr |> numeric_version()
#> [1] '2.0'
version_pkg |> numeric_version()
#> [1] '2.0'

version_dbl |> as.character() |> numeric_version()
#> [1] '2'
version_chr |> as.character() |> numeric_version()
#> [1] '2.0'
version_pkg |> as.character() |> numeric_version()
#> [1] '2.0'

version_dbl |> as.package_version()
#> Error: invalid version specification '2'
version_chr |> as.package_version()
#> [1] '2.0'
version_pkg |> as.package_version()
#> [1] '2.0'

version_dbl |> as.character() |> as.package_version()
#> Error: invalid version specification '2'
version_chr |> as.character() |> as.package_version()
#> [1] '2.0'
version_pkg |> as.character() |> as.package_version()
#> [1] '2.0'
```

gadenbuie commented:
Oh wait, no that wasn't quite right, we can't just use `numeric_version()` for the coercion because that's part of the problem to begin with.

I updated the logic to warn if `version` isn't a character string or a package_version, or numeric_version object and then temporarily set the `OutDec` option to `.` so that the character coercion happens as expected. 

The warning is included to help us track down incorrect usage and update to the character version.

wch commented:
I want to make sure I understand the problem correctly. Is it that, if someone using R-devel calls `is_installed()` with a number (instead of a string), it will raise this warning?

```
> is_installed("shiny", 1.7)
Warning in .make_numeric_version(x, strict, .standard_regexps()$valid_numeric_version) :
  invalid non-character version specification 'x' (type: double)
[1] TRUE
```

wch commented on inst/staticexports/package.R:
```diff
@@ -7,6 +7,14 @@ is_installed <- function(pkg, version = NULL) {
   if (is.null(version)) {
     return(installed)
   }
+  if (!inherits(version, c("package_version", "numeric_version"))) {
+    if (!is.character(version)) {
+      warning("`version` must be a character string or a `package_version` or `numeric_version` object.")
+      opts <- options(OutDec = ".")
+      on.exit(options(opts))
+      version <- numeric_version(as.character(version))
+    }
+  }
```
Some suggestions for this code:
- Use `is.character()` first, because it's faster, as @schloerke noted in #27. (Speed is a high priority for functions in staticimports.)
- Instead of nested `if` statements, use `&&`. (Putting `is.character()` first still has a speed advantage due to the short-circuit behavior of `&&`.)
- No need to check for `package_version` class, because those objects are also `numeric_version` objects.
- Use `sprintf()` to control formatting instead of setting the global `OutDec` option and calling `as.character()`.

```suggestion
  if (!is.character(version) && !inherits(version, "numeric_version")) {
    warning("`version` must be a character string or `numeric_version` object.")
    version <- numeric_version(sprintf("%0.9g", version))
  }
```

gadenbuie replied:
Refactored in d42f149e94bdeb1b69abec0560a4ee358bb4a0c4.

wch reviewed (changes requested):
In addition to the suggested changes, please also add unit tests (which could just be borrowed from #27).

gadenbuie commented:
> I want to make sure I understand the problem correctly. Is it that, if someone using R-devel calls `is_installed()` with a number (instead of a string), it will raise this warning?

Yes exactly, which is also the warning that prompted the patch release in Shiny. From internal conversations, this warning can be hard to track down but is easy to introduce, the most sneaky of which is the version comparison we have in `is_installed()`.

While rstudio/shiny#3850 fixed the problem in shiny, `is_installed()` is used throughout our packages. Because it's used in many places, I originally thought we should convert the version but with a warning so that we can locate non-character usage. In 1ce956d80f2e1d43ca258e5637037e4ea43f2a2e I modified this to throw an error in (testthat) testing but warn otherwise. I'm open to whatever behavior you want in the non-character case, but I was thinking the change in this PR would be a short-term bridge until we've found and converted all uses of `is_installed()`.

wch commented:
Looks good!
